File: go/internal/cli/collections.go
package cli

// Collections sub-commands — manage namespaces inside one chromem store.
//
//	houkai collections list           List collections with memory counts.
//	houkai collections create NAME    Create an empty collection.
//	houkai collections delete NAME    Delete a collection and its memories.
//	houkai collections copy SRC DST   Copy memories (with embeddings) SRC → DST.

import (
	"encoding/json"
	"fmt"
	"sort"

	"github.com/spf13/cobra"
)

func newCollectionsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "collections",
		Short: "Manage collections (namespaces) in the store",
	}
	cmd.AddCommand(
		newCollectionsListCmd(),
		newCollectionsCreateCmd(),
		newCollectionsDeleteCmd(),
		newCollectionsCopyCmd(),
	)
	return cmd
}

func newCollectionsListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list",
		Short: "List all collections in the store with their memory counts",
		RunE: func(cmd *cobra.Command, _ []string) error {
			backend := backendFromCtx(cmd.Context())
			active := cfgFromCtx(cmd.Context()).Collection

			counts := backend.ListCollections()
			names := make([]string, 0, len(counts))
			for n := range counts {
				names = append(names, n)
			}
			sort.Strings(names)

			if fmtFromCtx(cmd.Context()) == FormatJSON {
				type row struct {
					Name   string `json:"name"`
					Count  int    `json:"count"`
					Active bool   `json:"active"`
				}
				rows := make([]row, len(names))
				for i, n := range names {
					rows[i] = row{Name: n, Count: counts[n], Active: n == active}
				}
				b, _ := json.MarshalIndent(rows, "", "  ")
				fmt.Println(string(b))
				return nil
			}

			fmt.Printf("%-30s  %8s  %s\n", "COLLECTION", "MEMORIES", "")
			for _, n := range names {
				star := ""
				if n == active {
					star = "*"
				}
				fmt.Printf("%-30s  %8d  %s\n", n, counts[n], star)
			}
			return nil
		},
	}
}

func newCollectionsCreateCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "create <name>",
		Short: "Create an empty collection",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			backend := backendFromCtx(cmd.Context())
			name := args[0]
			if backend.HasCollection(name) {
				return fmt.Errorf("collection %q already exists", name)
			}
			if err := backend.CreateCollection(name); err != nil {
				return err
			}
			fmt.Printf("Created collection %q.\n", name)
			return nil
		},
	}
}

func newCollectionsDeleteCmd() *cobra.Command {
	var yes bool
	cmd := &cobra.Command{
		Use:   "delete <name>",
		Short: "Delete a collection and every memory in it (irreversible)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			backend := backendFromCtx(cmd.Context())
			name := args[0]
			if !backend.HasCollection(name) {
				return fmt.Errorf("collection %q not found", name)
			}
			if name == cfgFromCtx(cmd.Context()).Collection {
				return fmt.Errorf("refusing to delete the active collection %q — switch with --collection first", name)
			}
			count := backend.ListCollections()[name]
			if !yes && !Confirm(fmt.Sprintf("Delete collection %q (%d memories)?", name, count)) {
				fmt.Println("Aborted.")
				return nil
			}
			if err := backend.DeleteCollection(name); err != nil {
				return err
			}
			fmt.Printf("Deleted collection %q (%d memories).\n", name, count)
			return nil
		},
	}
	cmd.Flags().BoolVarP(&yes, "yes", "y", false, "Skip confirmation")
	return cmd
}

func newCollectionsCopyCmd() *cobra.Command {
	var yes bool
	cmd := &cobra.Command{
		Use:   "copy <src> <dst>",
		Short: "Copy all memories (embeddings included — no re-embedding) SRC → DST",
		Long: `Copy all memories (embeddings included — no re-embedding) SRC → DST.

DST is created if missing; existing DST ids are overwritten.`,
		Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			backend := backendFromCtx(cmd.Context())
			src, dst := args[0], args[1]
			if !backend.HasCollection(src) {
				return fmt.Errorf("collection %q not found", src)
			}
			if src == dst {
				return fmt.Errorf("SRC and DST are the same collection")
			}
			total := backend.ListCollections()[src]
			if total == 0 {
				fmt.Printf("Collection %q is empty — nothing to copy.\n", src)
				return nil
			}
			if !yes && !Confirm(fmt.Sprintf("Copy %d memories %q → %q?", total, src, dst)) {
				fmt.Println("Aborted.")
				return nil
			}
			copied, err := backend.CopyCollection(cmd.Context(), src, dst)
			if err != nil {
				return err
			}
			fmt.Printf("Copied %d memories %q → %q.\n", copied, src, dst)
			return nil
		},
	}
	cmd.Flags().BoolVarP(&yes, "yes", "y", false, "Skip confirmation")
	return cmd
}

File: ai_houkai/cli/commands/collections.py
"""Collections sub-commands — manage namespaces inside one Chroma store.

Commands
    houkai collections list           List collections with memory counts.
    houkai collections create NAME    Create an empty collection.
    houkai collections delete NAME    Delete a collection and its memories.
    houkai collections copy SRC DST   Copy memories (with embeddings) SRC → DST.
"""

from __future__ import annotations

import json
import sys

import typer
from rich.console import Console
from rich.table import Table

from ai_houkai.cli import output as out
from ai_houkai.memory_system.store import _get_embed_fn

collections_app = typer.Typer(
    name="collections",
    help="Manage collections (namespaces) in the store.",
    no_args_is_help=True,
)

_BATCH = 256


def _client(ctx: typer.Context):
    return ctx.obj["store"].client


def _open(ctx: typer.Context, name: str):
    """Open a collection the same way MemoryStore does (cosine + same EF)."""
    store = ctx.obj["store"]
    return store.client.get_or_create_collection(
        name=name,
        embedding_function=_get_embed_fn(store.embedding_model),
        metadata={"hnsw:space": "cosine"},
    )


@collections_app.command("list")
def list_cmd(
    ctx: typer.Context,
    fmt: str = typer.Option("auto", "--format", "-f", help="auto|json"),
) -> None:
    """List all collections in the store with their memory counts."""
    client = _client(ctx)
    active = ctx.obj["config"].collection
    rows = [
        {
            "name": col.name,
            "count": client.get_collection(col.name).count(),
            "active": col.name == active,
        }
        for col in sorted(client.list_collections(), key=lambda c: c.name)
    ]

    if fmt == "json":
        print(json.dumps(rows, indent=2))
        return
    if fmt == "auto" and not sys.stdout.isatty():
        for r in rows:
            print(f"{r['name']}\t{r['count']}\t{'*' if r['active'] else ''}")
        return
    t = Table(show_header=True, header_style="bold cyan")
    t.add_column("COLLECTION")
    t.add_column("MEMORIES", justify="right")
    t.add_column("", justify="center")
    for r in rows:
        t.add_row(r["name"], str(r["count"]), "*" if r["active"] else "")
    Console().print(t)


@collections_app.command("create")
def create_cmd(ctx: typer.Context, name: str) -> None:
    """Create an empty collection (cosine space, store's embedding model)."""
    client = _client(ctx)
    if any(c.name == name for c in client.list_collections()):
        typer.echo(f"Collection {name!r} already exists.")
        raise typer.Exit(1)
    _open(ctx, name)
    typer.echo(f"Created collection {name!r}.")


@collections_app.command("delete")
def delete_cmd(
    ctx: typer.Context,
    name: str,
    yes: bool = typer.Option(False, "--yes", "-y"),
) -> None:
    """Delete a collection and every memory in it. Irreversible."""
    client = _client(ctx)
    if not any(c.name == name for c in client.list_collections()):
        typer.echo(f"Collection {name!r} not found.", err=True)
        raise typer.Exit(1)
    if name == ctx.obj["config"].collection:
        typer.echo(
            f"Refusing to delete the active collection {name!r} — "
            "switch with --collection first.", err=True,
        )
        raise typer.Exit(1)
    count = client.get_collection(name).count()
    if not out.confirm(f"Delete collection {name!r} ({count} memories)?", yes=yes):
        typer.echo("Aborted.")
        return
    client.delete_collection(name)
    typer.echo(f"Deleted collection {name!r} ({count} memories).")


@collections_app.command("copy")
def copy_cmd(
    ctx: typer.Context,
    src: str,
    dst: str,
    yes: bool = typer.Option(False, "--yes", "-y"),
) -> None:
    """Copy all memories (embeddings included — no re-embedding) SRC → DST.

    DST is created if missing; existing DST ids are overwritten.
    """
    client = _client(ctx)
    if not any(c.name == src for c in client.list_collections()):
        typer.echo(f"Collection {src!r} not found.", err=True)
        raise typer.Exit(1)
    if src == dst:
        typer.echo("SRC and DST are the same collection.", err=True)
        raise typer.Exit(1)

    src_col = client.get_collection(src)
    total = src_col.count()
    if total == 0:
        typer.echo(f"Collection {src!r} is empty — nothing to copy.")
        return
    if not out.confirm(f"Copy {total} memories {src!r} → {dst!r}?", yes=yes):
        typer.echo("Aborted.")
        return

    dst_col = _open(ctx, dst)
    copied = 0
    for offset in range(0, total, _BATCH):
        res = src_col.get(
            include=["documents", "metadatas", "embeddings"],
            limit=_BATCH,
            offset=offset,
        )
        ids = res.get("ids") or []
        if not ids:
            break
        embs = res.get("embeddings")
        dst_col.upsert(
            ids=ids,
            documents=res.get("documents"),
            metadatas=res.get("metadatas"),
            embeddings=None if embs is None else [list(e) for e in embs],
        )
        copied += len(ids)
    typer.echo(f"Copied {copied} memories {src!r} → {dst!r}.")
